### DISCUSSION TITLE
Running MOOSE using apptainer / singularity

### DISCUSSION THREAD

**Discussion by cgentry7** [idaholab/moose#27377] [Q&A Getting Started]
I've been trying to test using MOOSE on our HPC system using Apptainer.  I've been following the instructions [here](https://mooseframework.inl.gov/getting_started/installation/docker.html), but can't get either the moose-dev image or the installed image to work with the instructions provided.

In the case of the installed image, I get a `Failed to copy the requested directory` error when attempting to copy `thermal_hydraulics`, and subsequently `Run failed` if I do try to run what does copy, with the following error messages:

```
Traceback (most recent call last):
  File "/opt/moose/share/moose/python/pyhit/__init__.py", line 18, in <module>
    import hit
ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /opt/moose/share/moose/python/hit.so)
```

For the moose-dev image, I get the follow errors when trying to make the test:

```
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/libtool: No such file or directory
/bin/sh: ./moose/libmesh/installed/libtool: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: ./moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
make: python-config: Command not found
make: python-config: Command not found
```

It looks like there are multiple things wrong with the environment configuration, with the `PATH` not including `/opt/moose/bin`, and seemingly missing libraries.  Maybe my expectation was wrong here, but as an image, I expected this to work more-so "out of the box".

Any help would be greatly appreciated.

**Comment by GiudGiud** [comment]
@loganharbour on apptainer

**Comment by loganharbour** [comment]
To make the docker image, we convert from apptainer -> docker. What you're doing now is then converting from docker -> apptainer again. Unfortunately, it looks like the environment sourcing is stripped from this double conversion.

I'm looking into getting that environment correct again. If I can't get that quickly, I'll at least find a way to share the apptainer sifs.

**Reply by cgentry7** [reply]
Awesome!  Yeah, when `/opt/moose/bin` wasn't in the `PATH`, I knew something had to be up with the environment configuration not being saved.  Thanks for the quick response!

**Reply by loganharbour** [reply]
The standard way that everyone does this is builds everything with docker and then just does the conversion from docker, which is a well defined process. We're stubborn and ended up starting the build with apptainer... and the conversion the other way around is still pretty simple but has a few issues :)

In the interim, this is the environment that _should_ be there (if you just put this in a file and source it, you should get the desired behavior)

```bash
    # Alternate GCC 12
    source /opt/rh/gcc-toolset-12/enable

# If discovered, use these files as initialization routines
if [ -f /init_env ]; then
  file="/init_env"
else
  file="/none"
fi
action="${0##*/}"
case "${action}" in
shell)
    if [ "${SINGULARITY_SHELL:-}" = "/bin/bash" ]; then
        set -- --noprofile --rcfile $file
    elif test -z "${SINGULARITY_SHELL:-}"; then
        export SINGULARITY_SHELL=/bin/bash
        set -- --noprofile --rcfile $file
    fi
    ;;
exec)
    export BASH_ENV="$file"
    set -- /bin/bash --noprofile --rcfile $file -c "$*"
    ;;
run)
    set -- /bin/bash --noprofile --rcfile $file
esac

    # Fix locale warnings
    export LC_ALL=C



    # Set the MPI environment
    source /opt/mpi/use-mpich

    # From moose-petsc
    export PETSC_DIR=/opt/petsc



    # From moose-libmesh
    export LIBMESH_DIR=/opt/libmesh
    if ! $CC --version | grep --quiet "clang version 16"; then
      export VTKINCLUDE_DIR=/opt/vtk/include/vtk-9.2
      export VTKLIB_DIR=/opt/vtk/lib
    fi



    export PATH=/opt/miniforge3/bin:$PATH
    source activate /opt/miniforge3/envs/moose

    export WASP_DIR=/opt/wasp
    export PATH=${WASP_DIR}/bin:$PATH


    export PATH=/opt/moose/bin:$PATH
```
